lib/src/main/java/tech/relaycorp/relaydroid/messaging/SendMessage.kt

```diff
@@ -0,0 +1,93 @@
+package tech.relaycorp.relaydroid.messaging
+
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import tech.relaycorp.poweb.PoWebClient
+import tech.relaycorp.relaydroid.PrivateThirdPartyEndpoint
+import tech.relaycorp.relaydroid.PublicThirdPartyEndpoint
+import tech.relaycorp.relaydroid.Relaynet
+import tech.relaycorp.relaydroid.RelaynetException
+import tech.relaycorp.relaydroid.common.Logging.logger
+import tech.relaycorp.relaynet.bindings.pdc.Signer
+import tech.relaycorp.relaynet.issueEndpointCertificate
+import tech.relaycorp.relaynet.messages.Parcel
+import tech.relaycorp.relaynet.wrappers.x509.Certificate
+import java.util.logging.Level
+import kotlin.coroutines.CoroutineContext
+
+internal class SendMessage(
+    private val poWebClientBuilder: () -> PoWebClient =
+        { PoWebClient.initLocal(Relaynet.POWEB_PORT) },
+    private val coroutineContext: CoroutineContext = Dispatchers.IO
+) {
+
+    suspend fun send(message: OutgoingMessage) {
+        withContext(coroutineContext) {
+            message.validate()
+
+            val parcel = buildParcel(message)
+            val senderPrivateKey = message.senderEndpoint.keyPair.private
+
+            return@withContext try {
+                poWebClientBuilder()
+                    .deliverParcel(
+                        parcel.serialize(senderPrivateKey),
+                        Signer(
+                            parcel.senderCertificate,
+                            senderPrivateKey
+                        )
+                    )
+            } catch (e: Exception) {
+                logger.log(Level.WARNING, "Error sending message", e)
+                throw SendMessageException("Could not deliver message to gateway", e)
+            }
+        }
+    }
+
+    private suspend fun buildParcel(message: OutgoingMessage) = Parcel(
+        recipientAddress = message.receiverEndpoint.address,
+        payload = message.message,
+        senderCertificate = getSenderCertificate(message),
+        messageId = message.id.value,
+        creationDate = message.creationDate,
+        ttl = message.ttl,
+        senderCertificateChain = getSenderCertificateChain(message)
+    )
```

Comment:
I don't think building the parcel and its required fields should be the responsibility of this class -- I think it should be the responsibility of the `OutgoingMessage`.